{
  "TeamMentor_Article": {
    "$": {
      "Metadata_Hash": "1840165463",
      "Content_Hash": "1536159090"
    },
    "Metadata": [
      {
        "Id": [
          "d0a056d3-aaed-4ccc-8b8c-e88da2df5e84"
        ],
        "Id_History": [
          "d0a056d3-aaed-4ccc-8b8c-e88da2df5e84,27a773a3-5b01-4f97-8d06-5af5407d7ec4,"
        ],
        "Library_Id": [
          "be5273b1-d682-4361-99d9-6234f2d47eb7"
        ],
        "Title": [
          "Where Exceptions Aren't Possible, Return Error Codes"
        ],
        "Category": [
          "Error and Exception Management"
        ],
        "Phase": [
          "Implementation"
        ],
        "Technology": [
          "C++"
        ],
        "Type": [
          "Guideline"
        ],
        "DirectLink": [
          "Where Exceptions Aren't Possible, Return Error Codes"
        ],
        "Tag": [
          ""
        ],
        "Security_Demand": [
          ""
        ],
        "Author": [
          ""
        ],
        "Priority": [
          ""
        ],
        "Status": [
          ""
        ],
        "Source": [
          "SI"
        ]
      }
    ],
    "Content": [
      {
        "$": {
          "Sanitized": "true",
          "DataType": "Html"
        },
        "Data": [
          "<h1>What to Do</h1>\r\n  <p>Whenever possible you should use an exception management strategy for error handling. If this is not possible due to compiler or language limitations then use error return codes when necessary.</p>\r\n  <h1>Why</h1>\r\n  <p>If your function has an error that can impact your caller and you do not notify the caller they can do nothing to handle this problem. The result could be an application crash or other denial of service problem. In an extreme condition it could result in erroneous application logic that may be exploited by a determined attacker.</p>\r\n  <h1>When</h1>\r\n  <p>If your function can fail in a way that impacts the logic of your caller and can have a detrimental effect on the running of your application, return an error code to notify the caller.</p>\r\n  <h1>How</h1>\r\n  <p>Check for error conditions in your code and return an error code to your caller when necessary. It is very common for an integer return value of -1 to be an error code. It is also common to use a bool return value and return false in the event of an error. You may also use SetLastError in Win32 if you need to specify additional error details.</p>\r\n  <h1>Problem Example</h1>\r\n  <p>In the example below the function does not check for an error and does not return an error code in the case of failure of the read call.</p>\r\n  <pre>bool ReadFromSocket(SOCKET socket, char &#42;recvBuffer, int len)<br />&#123;<br />  //first read from the socket<br />&nbsp; read(socket, recvBuffer, len, 0);<br /><br /><br />  //now parse the buffer<br />  return parseBuffer(recvBuffer, len);<br />&#125;</pre>\r\n  <h1>Solution Example</h1>\r\n  <p>In the following example, the read API is called and then the return value is checked for errors. The function will return an error code if the read call fails.</p>\r\n  <pre>bool ReadFromSocket(SOCKET socket, char &#42;recvBuffer, int len)<br />&#123;<br />  //first read from the socket<br />&nbsp; int ret = read(socket, recvBuffer, len, 0);<br /><br /><br />  if (ret &gt;= 0)<br />  &#123;<br />      //now parse the buffer<br />      return parseBuffer(recvBuffer, len);<br />  &#125;<br />  else<br />  &#123;<br />      //can call WSAGetLastError if we need error details<br />      return false;<br />  &#125;<br />&#125;</pre>\r\n  <h1>Related Items</h1>\r\n  <ul>\r\n    <li>\r\n      <a href=\"/article/72900050-60e7-4562-9812-2f80bf0310bb\">C&#43;&#43; Exceptions are used where possible</a>\r\n      <li>\r\n        <a href=\"/article/e4a3052e-13f2-4062-9a83-49185ff14b0a\">Centralize Error Handling</a>\r\n        <br />\r\n      </li>\r\n    </li>\r\n  </ul>"
        ]
      }
    ]
  }
}